### PR TITLE
schema : make update to partition tables when 'set tiflash replica'

### LINKED ISSUE #5266: Fix set tiflash replica schema change not apply to the partition tables
## Enhancement

Partition tables are not updated when the schema change to set tiflash replica. 
Thus we need to fix it.

### PR DESCRIPTION
**Pull request by hongyunyan**
### What problem does this PR solve?

Issue Number: close #5266 

Problem Summary:

### What is changed and how it works?

update the tiflash replica info to partition tables when 'set tiflash replica'.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp [hunk 1]
```diff
@@ -1200,21 +1200,38 @@ void SchemaBuilder<Getter, NameMapper>::applySetTiFlashReplica(TiDB::DBInfoPtr d
...
+            auto new_table_info = latest_table_info->producePartitionTableInfo(part_def.id, name_mapper);
+            applySetTiFlashReplica(db_info, new_table_info);
+        }
+    }
+
+    applySetTiFlashReplica(db_info, latest_table_info);
+}
+
+template <typename Getter, typename NameMapper>
+void SchemaBuilder<Getter, NameMapper>::applySetTiFlashReplica(TiDB::DBInfoPtr db_info, TiDB::TableInfoPtr table_info)
```
Comment (JaySon-Huang) [inline dbms/src/TiDB/Schema/SchemaBuilder.cpp:1217]: Rename this function to `applySetTiFlashReplicaPhysicalTable`?

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp [hunk 2]
```diff
@@ -1200,21 +1200,38 @@ void SchemaBuilder<Getter, NameMapper>::applySetTiFlashReplica(TiDB::DBInfoPtr d
...
-        throw Exception(fmt::format("{} is not a ManageableStorage", name_mapper.debugCanonicalName(*db_info, *latest_table_info)));
+        throw Exception(fmt::format("{} is not a ManageableStorage", name_mapper.debugCanonicalName(*db_info, *table_info)));
 
-    applySetTiFlashReplica(db_info, latest_table_info, managed_storage);
+    applySetTiFlashReplica(db_info, table_info, managed_storage);
 }
 
+
 template <typename Getter, typename NameMapper>
 void SchemaBuilder<Getter, NameMapper>::applySetTiFlashReplica(
```
Comment (JaySon-Huang) [inline dbms/src/TiDB/Schema/SchemaBuilder.cpp:1236]: change this function name to `applySetTiFlashReplicaPhysicalTable ` too

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp [hunk 3]
```diff
@@ -1200,21 +1200,38 @@ void SchemaBuilder<Getter, NameMapper>::applySetTiFlashReplica(TiDB::DBInfoPtr d
     {
         throw TiFlashException(fmt::format("miss table in TiKV : {}", table_id), Errors::DDL::StaleSchema);
     }
+
+    if (latest_table_info->isLogicalPartitionTable())
+    {
+        for (const auto & part_def : latest_table_info->partition.definitions)
+        {
+            auto new_table_info = latest_table_info->producePartitionTableInfo(part_def.id, name_mapper);
+            applySetTiFlashReplica(db_info, new_table_info);
```
Comment (JaySon-Huang) [inline dbms/src/TiDB/Schema/SchemaBuilder.cpp:1209]:
```suggestion
            auto new_part_table_info = latest_table_info->producePartitionTableInfo(part_def.id, name_mapper);
            applySetTiFlashReplica(db_info, new_part_table_info);
```

**Review by JaySon-Huang — APPROVED** [review]
LGTM

File: dbms/src/TiDB/Schema/SchemaBuilder.h [hunk 4]
```diff
@@ -87,7 +87,8 @@ struct SchemaBuilder
     void applyExchangeTablePartition(const SchemaDiff & diff);
 
     void applySetTiFlashReplica(TiDB::DBInfoPtr db_info, TableID table_id);
-    void applySetTiFlashReplica(TiDB::DBInfoPtr db_info, TiDB::TableInfoPtr table_info, ManageableStoragePtr storage);
+    void applySetTiFlashReplicaOnLogicalTable(TiDB::DBInfoPtr db_info, TiDB::TableInfoPtr table_info, ManageableStoragePtr storage);
+    void applySetTiFlashReplicaOnPhysicalTable(TiDB::DBInfoPtr db_info, TiDB::TableInfoPtr table_info, ManageableStoragePtr storage);
```
Comment (JaySon-Huang) [inline dbms/src/TiDB/Schema/SchemaBuilder.h:91]: Make db_info and table_info to be `const ... &` to avoid copying

**Comment by hongyunyan** [comment]
cc @breezewish

**Review by breezewish — APPROVED** [review]

**Pull request merged**